components/brave_wallet/browser/eth_abi_decoder.h

```diff
@@ -0,0 +1,26 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_ETH_ABI_DECODER_H_
+#define BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_ETH_ABI_DECODER_H_
+
+#include <string>
+#include <vector>
+#include "base/values.h"
+
+namespace brave_wallet {
+
+bool UniswapEncodedPathDecode(const std::string& encoded_path,
+                              // NOLINTNEXTLINE(runtime/references)
+                              std::vector<std::string>& path);
+
+bool ABIDecode(const std::vector<std::string>& types,
+               const std::string& data,
```

Comment:
This interface shouldn't take hex string, it should be `const std::vector<uint8_t>&`.
Before this PR, `GetTransactionInfoFromData` was doing simple info extraction, now we have dedicated decoder to extract complex data fields, we should do byte level operations instead of unnecessary string operations with lots of copying like substr and assignment. And we should only covert the byte array into hex string for the final results.

Reply:
Do you suggest this change just at the `ABIDecode` level or also the `GetTransactionInfoFromData` level? I wanted to restrict changes only to the `eth_abi_decoder.cc` and `eth_data_parser.cc` files, but open to a wider change too if you recommend that.

By the way, the copying and assignments are not that big of a problem due to use of `StringPiece` but I agree byte arrays are a more natural choice here, especially with the use of offsets.

Reply:
From `GetTransactionInfoFromData` all the way to `ABIDecode` should all be byte array.
And by copying here I meant every time you call `substr`, it "copy" part of the string to a newly constructed string. The use of `StringPiece` just mitigates the copy assignment from the result.

Reply:
Fixed here: cdd2758deee18228fe7b77efbc446ab04cb6396e